Refactor invoice behaviors into the extpreimage package

treygriffith opened the pull request:
In preparation for retrieving preimages in the witness beacon, this change moves the logic for retrieving preimages on invoices to the `extpreimage` package. It also moves some of the unique handling for getting payment hashes on invoices to the `extpreimage` package as well.

Normally, I would move these to the `channeldb` package, but this refactor also helps with the maintenance of our fork by decreasing our diff with upstream lnd.

Compare:
- [current diff](https://github.com/sparkswap/lnd/compare/b07499f227bd78dbceaa8c6ee047f1c033716cdf..k%23epic/cross-chain-preimage)
- [diff after this pr](https://github.com/sparkswap/lnd/compare/b07499f227bd78dbceaa8c6ee047f1c033716cdf..k%23fix/external-preimage-witness-beacon)

### To do
- [x] tests

treygriffith committed:
move external preimage invoice methods into the extpreimage package

  - add invoice struct to mirror the channeldb invoice
  - add methods to GetPaymentHash and GetPaymentPreimage on invoices
  - add tests for added methods

dannypaz commented on extpreimage/invoice.go:
```diff
@@ -0,0 +1,124 @@
+	switch {
+	// if there is a local preimage available, we should use it to settle the
+	// invoice
+	case !bytes.Equal(i.PaymentPreimage[:], zeroPreimage[:]):
```
My initial reaction to adding the `switch` statement inside of `GetPaymentPreimage` is that we should leave the original code (The statement `preimage = invoice.Terms`) in `link.go` and detect if there is an external preimage... and then call out to this method.

**link.go**
```
switch {
        ....
	case !bytes.Equal(invoice.Terms.PaymentPreimage[:], zeroPreimage[:]):
		preimage = invoice.Terms.PaymentPreimage
	}
	// if this is an invoice with an external preimage, we should retrieve it.
	case invoice.Terms.ExternalPreimage:
                // make the call to extpreimage.GetPyamentPreimage
```

Reasoning for this is we dont want the `extpreimage` service to be responsible for ALL preimages (grabbing the preimage for all invoices), instead we would only want it in the case of identifying that there is an external preimage.

If we wanted any shared logic to determine if an Invoice is external or not, we should add those properties to `Invoices.go` itself.

We still have a problem where every time we would need a preimage, we would have conditional statements.

Anyways, with that said. What are your thoughts on the stuff above?

treygriffith replied:
In general, I agree. However, I moved the logic here so as to keep our diff with the upstream small. Everything in this file really belongs in invoices.go, but I put it here to make maintenance easier.

dannypaz replied:
@treygriffith Moving the ext preimages to an `invoice_preimage` module is a great change. From our talk, we decided against having the extpreimage package hold the responsibility of ALL invoice preimages. We've created another module to handle this which still allows maintenance to be easier (its a new file).

dannypaz commented on extpreimage/invoice_test.go:
```diff
@@ -0,0 +1,394 @@
+package extpreimage_test
```
@treygriffith really really happy with these tests. +1

treygriffith committed:
move extpreimage invoice to channeldb

This moves the extpreimage invoice handling to the channeldb package,
where it really belongs, but in a separate file to still make
ongoing maintenance easier.

dannypaz commented on channeldb/invoice_preimages.go:
```diff
@@ -0,0 +1,99 @@
+// either by pulling it directly from the invoice, or by retrieving
+// it from the external preimage service if it is an external preimage
+// invoice.
+func (c *ContractTerm) GetPaymentPreimage(timeLock uint32, currentHeight uint32,
```
Is there a way to document the return values to that we can say (preimage, temp error, perm error)

dannypaz commented on channeldb/invoice_preimages.go:
```diff
@@ -0,0 +1,99 @@
+    err := registry.AddInvoicePreimage(invoiceHash, preimage)
+
+    if err != nil {
+      return zeroPreimage, err, nil
```
Just so we know, is there any action the user can perform if we fail to save the preimage?

treygriffith replied:
When a temporary error (like this one) is returned, the link will fail (become inactive). AFAIK, LND will not try to revive that link until the daemon is restarted.

In the witness beacon (#24 and #25) we'll retry these errors periodically, and eventually I think we should have some logic to retry links that have failed because of temporary preimage errors, but I think that's best saved for a later date.

Practically speaking, the user should restart their daemon if they encounter an issue like this.

dannypaz reviewed (approved):
Looks fine @treygriffith

treygriffith committed:
add temporary and permanent error types to better document return values

dannypaz reviewed (approved):
@treygriffith thanks for adding the types. LGTM
